MathLib/LinAlg/SimpleMatrixVectorProvider-impl.h

```diff
@@ -0,0 +1,281 @@
+getMatrix(MatrixSpecificationsProvider const& msp)
+{
+    std::size_t id = 0u;
+    auto const mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<false>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(MatrixSpecificationsProvider const& msp, std::size_t& id)
+{
+    auto mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<true>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(Matrix const& A)
+{
+    std::size_t id = 0u;
+    auto const& res = getMatrix_<false>(id, A);
+    if (!res.second) // no new object has been created
+        BLAS::copy(A, *res.first);
+    return *res.first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(Matrix const& A, std::size_t& id)
+{
+    auto const& res = getMatrix_<false>(id, A);
+    if (!res.second) // no new object has been created
+        BLAS::copy(A, *res.first);
+    return *res.first;
+}
+
+template<typename Matrix, typename Vector>
+void
+SimpleMatrixVectorProvider<Matrix, Vector>::
+releaseMatrix(Matrix const& A)
+{
+    auto it = _used_matrices.find(const_cast<Matrix*>(&A));
+    if (it == _used_matrices.end()) {
+        ERR("The given matrix has not been found. Cannot release it. Aborting.");
+        std::abort();
+    } else {
+        ::detail::transfer(_used_matrices, _unused_matrices, it);
+    }
+}
+
+template<typename Matrix, typename Vector>
+template<bool do_search, typename... Args>
+std::pair<Vector*, bool>
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getVector_(std::size_t& id, Args&&... args)
+{
+    return get_<do_search>(id, _unused_vectors, _used_vectors, std::forward<Args>(args)...);
+}
+
+
+template<typename Matrix, typename Vector>
+Vector&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getVector()
+{
+    std::size_t id = 0u;
+    return *getVector_<false>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Vector&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getVector(std::size_t& id)
+{
+    return *getVector_<true>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Vector&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getVector(MatrixSpecificationsProvider const& msp)
+{
+    std::size_t id = 0u;
+    auto const mat_spec = msp.getMatrixSpecifications();
+    return *getVector_<false>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Vector&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getVector(MatrixSpecificationsProvider const& msp, std::size_t& id)
+{
+    auto mat_spec = msp.getMatrixSpecifications();
```

Comment:
If my understanding is correct the `mat_spec` contains matrix/vector meta data. So `getVector()` should return a vector with the size specified in `mat_spec`. Why is this specification not const? For me the comment 2 lines below also indicates that `mat_spec` should be `const`?!